### ISSUE TITLE
Performance issue in day interval with "setFullYear"

### ISSUE THREAD

**Issue by alemarre** [mbostock/d3#2601]
Hi,
I have a big performance issue with Internet Explorer 11.
I narrowed it down and it appears to come from this function (https://github.com/mbostock/d3/blob/master/src/time/day.js) : 

```javascript
d3_time.day = d3_time_interval(function(date) {
    var day = new d3_date(2e3, 0);
    day.setFullYear(date.getFullYear(), date.getMonth(), date.getDate());
    return day;
  },
```
In my case, IE profiler spends more than __45 seconds__ in the `setFullYear` function at page startup, when it only need some hundreds milliseconds to achieve the same in Chrome or FF.

__It appears that the combination of `new Date(2000, 0)` followed by `setFullYear` is really slow.__

I set up this interesting test case to show it.
It compares performances related to the Date constructor used, followed by a call to `setFullYear`, using those constructors : 
- new Date()
- new Date(2e3)
- new Date(2e3, 0) (the one used in d3.js) 
- new Date(2e3, 0, 0)
- new Date(2e3, 0, 1)

```html
<html>
	<body>
		<script>
			testPerf();
			testPerf(2e3);
			testPerf(2e3, 0);
			testPerf(2e3, 0, 0);
			testPerf(2e3, 0, 1);

			function testPerf(year, month, day) {
				var start = new Date();
				for (var i=0; i<=10000; i++) {
					// call the right Date constructor given testPerf arguments
					var date = isNaN(year) ? new Date() : (isNaN(month) ? new Date(year) : (isNaN(day) ? new Date(year, month) : new Date(year, month, day)));
					// Change it by any date
					date.setFullYear(2012, 6, 15);
				}
				var end = new Date();
				console.log("new Date(" + year + ", " + month + ", " + day +") elapsedTime : " + (end-start) + "ms");
			}

		</script>
	</body>
</html>
```

Results on IE : 
> new Date(undefined, undefined, undefined) elapsedTime : 26ms
> new Date(2000, undefined, undefined) elapsedTime : 16ms
> new Date(2000, 0, undefined) elapsedTime : __1169ms__
> new Date(2000, 0, 0) elapsedTime : 41ms
> new Date(2000, 0, 1) elapsedTime : __1135ms__

Results on Chrome (equivalent to FF) : 
> new Date(undefined, undefined, undefined) elapsedTime : 20ms
> new Date(2000, undefined, undefined) elapsedTime : 15ms
> new Date(2000, 0, undefined) elapsedTime : 11ms
> new Date(2000, 0, 0) elapsedTime : 10ms
> new Date(2000, 0, 1) elapsedTime : 9ms

I changed this line, and had my performances back to normal in IE :
```javascript
    var day = new d3_date(2e3, 0, 0);
```

What do you think of it ? Can this modification lead to other issues ?
Should I propose a pull request ?

Thanks

**Comment by mbostock** [comment]
I expect this issue has been fixed in 4.0’s [d3-time module](https://github.com/d3/d3-time), which uses a simpler strategy for computing the day boundary. The simpler strategy was buggy on older browsers (see #1197 d3/d3-time#1) but this has mostly been fixed, and the strategy used by D3 3.x wasn’t perfect anyway. The main improvement is in how the interval is implemented on top of floor and offset.

It’s rather unfortunate that IE11 performs this way. I don’t think it’s feasible to backport the 4.0 implementation to 3.x. My guess is that switching to `new Date(2000, 0, 0)` would be safe, but anything involving calendars is complicated and so it’s quite possible this could have an unintended side-effect in some obscure case.

**Comment by mbostock** [comment]
Closing as this code has been rewritten for 4.0.
